File: engines/waar-cohort/src/Game/Random/ConsequenceSampler.php
<?php

namespace App\Game\Random;

/** Dedicated binomial stream. Protocol and numerical method: specification §9.5. */
final class ConsequenceSampler
{
    public const VERSION = 'sha256-counter52-binomial-btrs/1';
    private int $counter = 0;
    private string $domain;

    public function __construct(int $seed, string $side, string $type, string $stage)
    {
        $this->domain = implode("\0", [self::VERSION, 'wounded-capture-then-compress/3', (string)$seed, $side, $type, $stage]);
    }

    /** Open interval (0,1), 52 random bits, no zero/log(0) or endpoint retry. */
    public function uniform(): float
    {
        $words = unpack('Nhigh/Nlow', substr(hash('sha256', $this->domain."\0".$this->counter++, true), 0, 8));
        return ($words['high'] * 1048576 + ($words['low'] >> 12) + 0.5) / 4503599627370496;
    }

    public function binomial(int $n, int $percent): int
    {
        if ($n < 0 || $n > 4294967295 || $percent < 0 || $percent > 100) throw new \InvalidArgumentException('Invalid consequence binomial parameters.');
        if ($n === 0 || $percent === 0) return 0;
        if ($percent === 100) return $n;
        // Compute the smaller probability from integer percentages in both runtimes.
        $p = min($percent, 100 - $percent) / 100;
        if ($n * $p < 30) {
            // Invert geometric waiting times; expected work <= 31 uniforms, even for u32 n.
            $limit = log1p(-$p); $position = 0.0; $sample = 0;
            while (true) {
                $position += floor(log($this->uniform()) / $limit) + 1;
                if ($position > $n) break;
                ++$sample;
            }
        } else {
            $sample = $this->btrs($n, $p);
        }
        return $percent > 50 ? $n - $sample : $sample;
    }

    /** Hörmann's transformed rejection, with the logarithmic acceptance test. */
    private function btrs(int $n, float $p): int
    {
        $sigma = sqrt($n * $p * (1 - $p));
        $b = 1.15 + 2.53 * $sigma;
        $a = -0.0873 + 0.0248 * $b + 0.01 * $p;
        $center = $n * $p + 0.5;
        $squeeze = 0.92 - 4.2 / $b;
        $alpha = (2.83 + 5.1 / $b) * $sigma;
        $mode = floor(($n + 1) * $p);
        $odds = $p / (1 - $p);
        while (true) {
            $u = $this->uniform() - 0.5; $v = $this->uniform();
            $us = 0.5 - abs($u);
            $k = floor((2 * $a / $us + $b) * $u + $center);
            if ($k < 0 || $k > $n) continue;
            if ($us >= 0.07 && $v <= $squeeze) return (int)$k;
            // log(P(X=k)/P(X=mode)); log1p avoids near-one cancellation at large n.
            $bound = ($mode + 0.5) * log(($mode + 1) / ($odds * ($n - $mode + 1)))
                + ($n + 1) * log1p(($k - $mode) / ($n - $k + 1))
                + ($k + 0.5) * log($odds * ($n - $k + 1) / ($k + 1))
                + self::tail($mode) + self::tail($n - $mode) - self::tail($k) - self::tail($n - $k);
            if (log($v * $alpha / ($a / ($us * $us) + $b)) <= $bound) return (int)$k;
        }
    }

    /** log(k!) - ((k+.5)log(k+1)-(k+1)+log(2π)/2). */
    private static function tail(float $k): float
    {
        $small = [0.08106146679532726, 0.04134069595540929, 0.02767792568499834, 0.02079067210376509,
            0.01664469118982119, 0.01387612882307075, 0.01189670994589177, 0.01041126526197210,
            0.009255462182712733, 0.008330563433362871];
        if ($k < 10) return $small[(int)$k];
        $x = $k + 1; $square = $x * $x;
        return (1 / 12 - (1 / 360 - (1 / 1260 - (1 / 1680 - 1 / 1188 / $square) / $square) / $square) / $square) / $x;
    }
}

File: engines/waar-cohort/rust/src/consequence_sampler.rs
//! Dedicated consequence stream; see specification §9.5. Never used by physics.
use sha2::{Digest, Sha256};

pub const VERSION: &str = "sha256-counter52-binomial-btrs/1";

pub struct ConsequenceSampler {
    domain: String,
    counter: u64,
}

impl ConsequenceSampler {
    pub fn new(seed: i64, side: &str, unit: &str, stage: &str) -> Self {
        Self {
            domain: format!(
                "{VERSION}\0wounded-capture-then-compress/3\0{seed}\0{side}\0{unit}\0{stage}"
            ),
            counter: 0,
        }
    }

    pub fn uniform(&mut self) -> f64 {
        let hash = Sha256::digest(format!("{}\0{}", self.domain, self.counter).as_bytes());
        self.counter += 1;
        let bits = u64::from_be_bytes(hash[..8].try_into().unwrap()) >> 12;
        (bits as f64 + 0.5) / 4503599627370496.0
    }

    pub fn binomial(&mut self, n: u32, percent: u32) -> u32 {
        assert!(percent <= 100);
        if n == 0 || percent == 0 {
            return 0;
        }
        if percent == 100 {
            return n;
        }
        let p = percent.min(100 - percent) as f64 / 100.0;
        let sample = if (n as f64) * p < 30.0 {
            let limit = (-p).ln_1p();
            let mut position = 0.0;
            let mut sample = 0;
            loop {
                position += (self.uniform().ln() / limit).floor() + 1.0;
                if position > n as f64 {
                    break;
                }
                sample += 1;
            }
            sample
        } else {
            self.btrs(n, p)
        };
        if percent > 50 {
            n - sample
        } else {
            sample
        }
    }

    fn btrs(&mut self, count: u32, p: f64) -> u32 {
        let n = count as f64;
        let sigma = (n * p * (1.0 - p)).sqrt();
        let b = 1.15 + 2.53 * sigma;
        let a = -0.0873 + 0.0248 * b + 0.01 * p;
        let center = n * p + 0.5;
        let squeeze = 0.92 - 4.2 / b;
        let alpha = (2.83 + 5.1 / b) * sigma;
        let mode = ((n + 1.0) * p).floor();
        let odds = p / (1.0 - p);
        loop {
            let u = self.uniform() - 0.5;
            let v = self.uniform();
            let us = 0.5 - u.abs();
            let k = ((2.0 * a / us + b) * u + center).floor();
            if k < 0.0 || k > n {
                continue;
            }
            if us >= 0.07 && v <= squeeze {
                return k as u32;
            }
            let bound = (mode + 0.5) * ((mode + 1.0) / (odds * (n - mode + 1.0))).ln()
                + (n + 1.0) * ((k - mode) / (n - k + 1.0)).ln_1p()
                + (k + 0.5) * (odds * (n - k + 1.0) / (k + 1.0)).ln()
                + tail(mode)
                + tail(n - mode)
                - tail(k)
                - tail(n - k);
            if (v * alpha / (a / (us * us) + b)).ln() <= bound {
                return k as u32;
            }
        }
    }
}

fn tail(k: f64) -> f64 {
    let small = [
        0.08106146679532726,
        0.04134069595540929,
        0.02767792568499834,
        0.02079067210376509,
        0.01664469118982119,
        0.01387612882307075,
        0.01189670994589177,
        0.01041126526197210,
        0.009255462182712733,
        0.008330563433362871,
    ];
    if k < 10.0 {
        return small[k as usize];
    }
    let x = k + 1.0;
    let square = x * x;
    (1.0 / 12.0
        - (1.0 / 360.0 - (1.0 / 1260.0 - (1.0 / 1680.0 - 1.0 / 1188.0 / square) / square) / square)
            / square)
        / x
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn consequence_vectors() {
        let vectors: serde_json::Value = serde_json::from_str(include_str!(
            "../../tests/fixtures/consequence-v3-vectors.json"
        ))
        .unwrap();
        for row in vectors["samples"].as_array().unwrap() {
            let mut sampler = ConsequenceSampler::new(
                row["seed"].as_i64().unwrap(),
                row["side"].as_str().unwrap(),
                row["type"].as_str().unwrap(),
                row["stage"].as_str().unwrap(),
            );
            let actual = sampler.binomial(
                row["n"].as_u64().unwrap() as u32,
                row["percent"].as_u64().unwrap() as u32,
            );
            assert_eq!(actual as u64, row["sample"].as_u64().unwrap(), "{row}");
            assert_eq!(
                (sampler.uniform() * 4503599627370496.0).floor() as u64,
                row["nextBits"].as_u64().unwrap(),
                "stream consumption: {row}"
            );
        }
    }
}
